### PR TITLE
Implement SxxExx EpisodeNum Processing

### LINKED ISSUE #1008: SxxExx Is not implemented causing guide information to be ignored
**Describe the bug**
The SxxExx episodenum system is not implemented. Guides information in this system from guides is ignored.

**To Reproduce**
https://github.com/jellyfin/jellyfin/blob/716586850962389ad8d7944f41341b7e8730b35c/Emby.XmlTv/Emby.XmlTv/Classes/XmlTvReader.cs#L500

**Expected behavior**
Episode season and episode numbers should be parsed and used from SxxExx system

**Logs**
N/A

**Screenshots**
N/A

**System (please complete the following information):**
 - Jellyfin Version: [e.g. 10.2.1]

### PR DESCRIPTION
**Pull request by daniel-widrick**
**Changes**
Implement and use SxxExx Episode numbering system from guide data if available.

**Issues**
Fixes #1008

### CODE REVIEW

**Comment by WWWesten** [comment]
Stupid question: Will this implementation correctly parse a episodes in the form SхххExxx, for example S099E321?

**Comment by daniel-widrick** [comment]
@WWWesten S099E321 should parse fine. 

99 - 321...

It will certainly throw an exception for something like S001E002b... Notably I am pretty sure Int.Parse("2b") will throw an exception that should probably be caught.. Best case that turns into 0.

File: Emby.XmlTv/Emby.XmlTv/Classes/XmlTvReader.cs [hunk 1]
```diff
@@ -495,16 +495,28 @@ public void ProcessEpisodeNum(XmlReader reader, XmlTvProgram result)
...
                     break;
             }
         }
+        public void ParseSxxExxSystem(XmlReader reader, XmlTvProgram result)
+        {
+
 
+            string value = reader.ReadElementContentAsString().ToUpper();
+            //Validate format
+            if (Regex.IsMatch(value,"[S][0-9]+[E][0-9]+"))
```
Comment (JustAMan) [inline Emby.XmlTv/Emby.XmlTv/Classes/XmlTvReader.cs]:
```suggestion
            if (Regex.IsMatch(value, "S[0-9]+E[0-9]+"))
```

File: Emby.XmlTv/Emby.XmlTv/Classes/XmlTvReader.cs [hunk 2]
```diff
@@ -495,16 +495,28 @@ public void ProcessEpisodeNum(XmlReader reader, XmlTvProgram result)
...
+            {
+                //Split at E - take before and remove S
```
Comment (JustAMan) [inline Emby.XmlTv/Emby.XmlTv/Classes/XmlTvReader.cs]: if you're using a regex, why not extract fields it already matched? it would handle cases like `S001E005foo` easily if you just take groups from `S([0-9]+)E([0-9]+)`

Comment (daniel-widrick) [reply Emby.XmlTv/Emby.XmlTv/Classes/XmlTvReader.cs]: Good call.. was [sS] before I decided to ToUpper() the value and dodge the issue. Forgot to remove the brackets.

Comment (Bond-009) [reply Emby.XmlTv/Emby.XmlTv/Classes/XmlTvReader.cs]: `[sS]` is better than `ToUpper` imo
Also you should use `ToUpperInvariant` ;)

Comment (daniel-widrick) [reply Emby.XmlTv/Emby.XmlTv/Classes/XmlTvReader.cs]: Invariant change in place.

I'm no regex master in csharp so having trouble getting Regex.Match to behave as expected.

			string input = "s002e45b";
			Match res = Regex.Match(input, "[sS]([0-9]+)[eE]([0-9]+)");

Doesn't seem to work as expected. An along those lines I don't know how far we want to "stretch" the SxxExx system. I can't actually find any documentation that it exists as a standard anywhere. How loose do we want to get? We quickly run into the territory of supporting things like:

Se002-Ep032
s2/ep32
s002xe32

Perhaps it's best to just take some form of the current match and log any non complying entries as "malformed"?

Comment (daniel-widrick) [reply Emby.XmlTv/Emby.XmlTv/Classes/XmlTvReader.cs]: Got the Regex.Match to work. Not sure if res.Groups[1].Success and res.Groups[2].Success should be check before using value or if the res.Success logically covers both

Comment (JustAMan) [reply Emby.XmlTv/Emby.XmlTv/Classes/XmlTvReader.cs]: I don't know for sure but my wild guess is `res.Sucess` should cover both (doesn't make much sense otherwise TBH).

File: Emby.XmlTv/Emby.XmlTv/Classes/XmlTvReader.cs [hunk 3]
```diff
@@ -495,16 +495,33 @@ public void ProcessEpisodeNum(XmlReader reader, XmlTvProgram result)
...
                     reader.Skip();
                     break;
             }
         }
+        public void ParseSxxExxSystem(XmlReader reader, XmlTvProgram result)
+        {
+            //// <episode-num system="SxxExx">S012E32</episode-num
+            string value = reader.ReadElementContentAsString();
+            Regex exp = new Regex("s([0-9]+)e([0-9]+)",RegexOptions.IgnoreCase);
+            Match res = exp.Match(value);
```
Comment (JustAMan) [inline Emby.XmlTv/Emby.XmlTv/Classes/XmlTvReader.cs]: can we combine two lines into
```suggestion
            Match res = new Regex("s([0-9]+)e([0-9]+)", RegexOptions.IgnoreCase).Match(value);
```
?

File: Emby.XmlTv/Emby.XmlTv/Classes/XmlTvReader.cs [hunk 4]
```diff
@@ -495,16 +495,33 @@ public void ProcessEpisodeNum(XmlReader reader, XmlTvProgram result)
...
+            if (res.Success)
+            {
+                int parsedInt;
 
+                if(int.TryParse(res.Groups[1].Value, out parsedInt))
```
Comment (JustAMan) [inline Emby.XmlTv/Emby.XmlTv/Classes/XmlTvReader.cs]:
```suggestion
                if (int.TryParse(res.Groups[1].Value, out parsedInt))
```

File: Emby.XmlTv/Emby.XmlTv/Classes/XmlTvReader.cs [hunk 5]
```diff
@@ -495,16 +495,33 @@ public void ProcessEpisodeNum(XmlReader reader, XmlTvProgram result)
...
+                {
+                    result.Episode.Series = parsedInt;
+                }
+                if(int.TryParse(res.Groups[2].Value, out parsedInt))
```
Comment (JustAMan) [inline Emby.XmlTv/Emby.XmlTv/Classes/XmlTvReader.cs]:
```suggestion
                if (int.TryParse(res.Groups[2].Value, out parsedInt))
```

File: Emby.XmlTv/Emby.XmlTv/Classes/XmlTvReader.cs [hunk 6]
```diff
@@ -495,16 +495,33 @@ public void ProcessEpisodeNum(XmlReader reader, XmlTvProgram result)
...
                     reader.Skip();
                     break;
             }
         }
+        public void ParseSxxExxSystem(XmlReader reader, XmlTvProgram result)
+        {
+            //// <episode-num system="SxxExx">S012E32</episode-num
+            string value = reader.ReadElementContentAsString();
+            Regex exp = new Regex("s([0-9]+)e([0-9]+)",RegexOptions.IgnoreCase);
+            Match res = new Regex("s([0-9]+)e([0-9]+)", RegexOptions.IgnoreCase).Match(value);
```
Comment (Bond-009) [inline Emby.XmlTv/Emby.XmlTv/Classes/XmlTvReader.cs]: Why did you change to a `Regex` instance instead of the static one?
Depending on the amount of times this function is executed this can have big performance consequences.
(Also, you are creating 2 of them :p)

Comment (daniel-widrick) [reply Emby.XmlTv/Emby.XmlTv/Classes/XmlTvReader.cs]: Sorry was going through the requested changes quickly yesterday.
https://github.com/jellyfin/jellyfin/pull/1009/commits/49f4be6e7949a9cd269980aa97792e7453baf6c6

Should address the issues.

File: Emby.XmlTv/Emby.XmlTv/Classes/XmlTvReader.cs [hunk 7]
```diff
@@ -495,16 +495,32 @@ public void ProcessEpisodeNum(XmlReader reader, XmlTvProgram result)
...
+            Match res = Regex.Match(value, "s([0-9]+)e([0-9]+)", RegexOptions.IgnoreCase);
+            if (res.Success)
+            {
+                int parsedInt;
 
+                if (int.TryParse(res.Groups[1].Value, out parsedInt))
+                {
+                    result.Episode.Series = parsedInt;
+                }
+                if (int.TryParse(res.Groups[2].Value, out parsedInt))
```
Comment (Bond-009) [inline Emby.XmlTv/Emby.XmlTv/Classes/XmlTvReader.cs]:
```suggestion

                if (int.TryParse(res.Groups[2].Value, out parsedInt))
```
Newline after closing brace.

Comment (daniel-widrick) [reply Emby.XmlTv/Emby.XmlTv/Classes/XmlTvReader.cs]: Is there a style guide published somewhere for this?

Comment (Bond-009) [reply Emby.XmlTv/Emby.XmlTv/Classes/XmlTvReader.cs]: Not yet, we do have some analyzers to enforce them.

Comment (daniel-widrick) [reply Emby.XmlTv/Emby.XmlTv/Classes/XmlTvReader.cs]: welll.. I believe the new function is in line with the rest of the file now.
